Allow range filters on columns with coercions

mbasmanova opened the pull request:
Coercion is now supported for columns with any type of filter and regardless of whether column is filter-only or is being projected out.

```
== NO RELEASE NOTE ==
```

oerling commented:
Nice structure. LGTM.

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Friday, November 8, 2019 7:24 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Review requested <review_requested@noreply.github.com>
Subject: [prestodb/presto] Allow range filters on columns with coercions (#13674)

 

The remaining work is to add support for filters on columns with coercions when these columns are only used for filtering and are not projected out. This change will be in a separate PR.

== NO RELEASE NOTE ==
  _____  


You can view, comment on, or merge this pull request online at:


  https://github.com/prestodb/presto/pull/13674


Commit Summary


*	Introduce HiveCoercer interface
*	Add coercing tuple domain filters
*	Static import TupleDomain.withColumnDomains
*	Allow range filters on columns with coercion


File Changes


*	R presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercer.java <https://github.com/prestodb/presto/pull/13674/files#diff-0>  (160) 
*	M presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java <https://github.com/prestodb/presto/pull/13674/files#diff-1>  (2) 
*	M presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java <https://github.com/prestodb/presto/pull/13674/files#diff-2>  (6) 
*	M presto-hive/src/main/java/com/facebook/presto/hive/HiveSelectivePageSourceFactory.java <https://github.com/prestodb/presto/pull/13674/files#diff-3>  (4) 
*	M presto-hive/src/main/java/com/facebook/presto/hive/orc/DwrfSelectivePageSourceFactory.java <https://github.com/prestodb/presto/pull/13674/files#diff-4>  (5) 
*	M presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java <https://github.com/prestodb/presto/pull/13674/files#diff-5>  (34) 
*	M presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java <https://github.com/prestodb/presto/pull/13674/files#diff-6>  (95) 
*	A presto-hive/src/test/java/com/facebook/presto/hive/TestCoercingFilters.java <https://github.com/prestodb/presto/pull/13674/files#diff-7>  (59) 
*	M presto-main/src/main/java/com/facebook/presto/testing/MaterializedResult.java <https://github.com/prestodb/presto/pull/13674/files#diff-8>  (3) 
*	M presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java <https://github.com/prestodb/presto/pull/13674/files#diff-9>  (1) 
*	M presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java <https://github.com/prestodb/presto/pull/13674/files#diff-10>  (2) 


Patch Links:


*	https://github.com/prestodb/presto/pull/13674.patch
*	https://github.com/prestodb/presto/pull/13674.diff

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13674?email_source=notifications&email_token=AKPPPTYYLYPDU7VSLCIZFI3QSV77VA5CNFSM4JKY4PG2YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4HX7WJ5A> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKPPPT3E5F32B5PKGOH7NJ3QSV77VANCNFSM4JKY4PGQ> .

mbasmanova commented:
@oerling Turns out that coercion works for filter-only columns with no changes. This is because `HiveSplit#partitionSchemaDifference` includes all the columns with different schema, not just the ones projected out of the scan, and this is the data that's used to generate coercions. Added a test and removed a TODO.

bhhari commented on presto-main/src/main/java/com/facebook/presto/testing/MaterializedResult.java:
```diff
@@ -419,6 +419,9 @@ public static MaterializedResult materializeSourceDataStream(ConnectorSession se
                 break;
             }
+            if (outputPage.getPositionCount() == 0) {
+                continue;
```
should this be also break and combine with above if

mbasmanova replied:
@bhhari `null` page means there is no more data; empty page means that all the data in a batch was filtered out, but there still might be more batches.

bhhari commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercer.java:
```diff
@@ -346,5 +468,15 @@ else if (i < rowBlock.getFieldCount()) {
+        @Override
+        public TupleDomainFilter toCoercingFilter(TupleDomainFilter filter)
+        {
+            if (filter == IS_NULL || filter == IS_NOT_NULL) {
```
what about ALWAYS_FALSE and ALWAYS_TRUE

mbasmanova replied:
@bhhari We should't have these. If there is an ALWAYS_FALSE filter, the optimizer should replace Scan with empty `Values`.

bhhari commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercer.java:
```diff
@@ -259,14 +361,24 @@ public Block apply(Block block)
+        @Override
+        public TupleDomainFilter toCoercingFilter(TupleDomainFilter filter)
+        {
+            if (filter == IS_NULL || filter == IS_NOT_NULL) {
```
ALWAYS_FALSE?

mbasmanova replied:
@bhhari See my answer above.

bhhari commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -1209,7 +1237,9 @@ protected void doTestMismatchSchemaTable(
             List<ColumnMetadata> tableAfter,
-            MaterializedResult dataAfter)
+            MaterializedResult dataAfter,
+            List<TupleDomain<ColumnMetadata>> afterFilters,
```
this requires a change in AbstractTestPrismClient in presto-facebook
